### PR TITLE
Fix data race between SSL_SESSION_list_add and ssl_session_dup

### LINKED ISSUE #24629: Data race in tls_process_new_session_ticket?
Hi all,

we ([ClickHouse](https://github.com/ClickHouse/ClickHouse), an open-source analytical database) recently migrated from boringssl to OpenSSL 3.2 ([PR](https://github.com/ClickHouse/ClickHouse/pull/59870)). Back then, session caching had to be disabled ([commit](https://github.com/ClickHouse/ClickHouse/pull/59870/commits/d30b48f54e42ec5aeb1bc27c937adb8efdb8108d) as part of aforementioned PR) as our continuous tests using [ThreadSanitizer](https://clang.llvm.org/docs/ThreadSanitizer.html) builds found data races in OpenSSL and the effort had to be completed in time back then.

I re-visited session caching today (i.e. re-enabled it), and looked more closely into the the races:
- [here](https://github.com/ClickHouse/ClickHouse/pull/65111) is my current PR (it is wip)
- [here](https://s3.amazonaws.com/clickhouse-test-reports/65111/0b7c30de12f9ea9cb9ad645a81fbe8de07352a4a/integration_tests__tsan__[1_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel0_0.log) is a report of the race, search "Sanitizer assert".

I am not an OpenSSL expert but I think this is what happens:

This thread adds a SSL session to the session cache. This operation is performed by `SSL_CTX_add_session` under an exclusive lock:

```
E Previous write of size 8 at 0x7260000d0b30 by thread T672 (mutexes: write M1, write M2):
E #0 SSL_SESSION_list_add build_docker/./contrib/openssl/ssl/ssl_sess.c:1263:27 (clickhouse+0x20250cea) (BuildId: a38098db2d267457b06b8d4a8b29e4d4d8b98a0a)
E #1 SSL_CTX_add_session build_docker/./contrib/openssl/ssl/ssl_sess.c:762:5 (clickhouse+0x20250cea)
E #2 ssl_update_cache build_docker/./contrib/openssl/ssl/ssl_lib.c:4527:13 (clickhouse+0x2024632c) (BuildId: a38098db2d267457b06b8d4a8b29e4d4d8b98a0a)
E #3 tls_process_new_session_ticket build_docker/./contrib/openssl/ssl/statem/statem_clnt.c:2850:9 (clickhouse+0x202c3ba6) (BuildId: a38098db2d267457b06b8d4a8b29e4d4d8b98a0a)
[...]
```

This thread wants to replace the existing cached session:

```
E WARNING: ThreadSanitizer: data race (pid=1)
E Read of size 8 at 0x7260000d0b30 by thread T674 (mutexes: write M0):
E #0 __tsan_memcpy <null> (clickhouse+0x7237feb) (BuildId: a38098db2d267457b06b8d4a8b29e4d4d8b98a0a)
E #1 ssl_session_dup build_docker/./contrib/openssl/ssl/ssl_sess.c:146:5 (clickhouse+0x2024f2f9) (BuildId: a38098db2d267457b06b8d4a8b29e4d4d8b98a0a)
E #2 tls_process_new_session_ticket build_docker/./contrib/openssl/ssl/statem/statem_clnt.c:2731:25 (clickhouse+0x202c354f) (BuildId: a38098db2d267457b06b8d4a8b29e4d4d8b98a0a)
E #3 ossl_statem_client_process_message build_docker/./contrib/openssl/ssl/statem/statem_clnt.c:1126:16 (clickhouse+0x202bf906) (BuildId: a38098db2d267457b06b8d4a8b29e4d4d8b98a0a)
[...]
```

`tls_process_new_session_ticket` has a comment `Sessions must be immutable once they go into the session cache. Otherwise we can get multi-thread problems. Therefore we don't "update" sessions, we replace them with a duplicate. In TLSv1.3 we need to do this every time a NewSessionTicket arrives because those messages arrive post-handshake and the session may have already gone into the session cache.`.

"Duplicating" the session happens via memcpy in `ssl_session_dup` without lock and I did not find any other mechanism further up the stack that would protect the cached session against parallel modification or release. I guess thread sanitizer complains rightfully here unless I missed something. The copy needs to happen under a read lock to be kosher.

Please tell me I missed something 😄.

### PR DESCRIPTION
**Pull request by rschu1ze**
Fixes #24629

This upstreams a fix for data race #24629 which has been made a while ago (https://github.com/openssl/openssl/issues/24629#issuecomment-2174295424)

It resolved the tsan warning in my local testing, nevertheless kindly double check if everything is kosher
(https://github.com/openssl/openssl/issues/24629#issuecomment-2174309613)

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: ssl/ssl_sess.c [hunk 1]
```diff
@@ -138,7 +138,15 @@ static SSL_SESSION *ssl_session_dup_intern(const SSL_SESSION *src, int ticket)
     dest = OPENSSL_malloc(sizeof(*dest));
     if (dest == NULL)
         return NULL;
-    memcpy(dest, src, sizeof(*dest));
+
+    /*
+     * Copy until prev ptr, because it's a part of sessons cache which can be modified
+     * concurrently. Other fields filled in the code bellow.
```
Comment (mattcaswell) [inline ssl/ssl_sess.c:144]: nit: typo "below"

File: ssl/ssl_sess.c [hunk 2]
```diff
@@ -138,7 +138,15 @@ static SSL_SESSION *ssl_session_dup_intern(const SSL_SESSION *src, int ticket)
...
+     */
+    memcpy(dest, src, offsetof(SSL_SESSION, prev));
+    dest->ext = src->ext;
+    dest->ticket_appdata_len = src->ticket_appdata_len;
+    dest->flags = src->flags;
```
Comment (mattcaswell) [inline ssl/ssl_sess.c:149]: These 3 lines would not be necessary if we moved the declaration of prev/next to be at the end of the structure.

**Review by mattcaswell — COMMENTED** [review]
I'd suggest we should also add a comment into the ssl_local.h around the definition of ssl_session_st pointing out that changing the members of the structure may require updates to ssl_session_dup_intern.

Additionally @davidben points out that the `references` member should probably also not be copied via memcpy (which can probably most easily be achieved by moving the declaration of that field *after* prev/next.

**Review by nhorman — COMMENTED** [review]
did we also want to consider moving the reference count to the bottom of the struct as you suggested @mattcaswell ?  I know this abates the race condition, but it seems like a good idea on principle

File: ssl/ssl_sess.c (hunk 1 above)
Comment (rschu1ze) [reply ssl/ssl_sess.c:144]: Done

File: ssl/ssl_sess.c (hunk 2 above)
Comment (rschu1ze) [reply ssl/ssl_sess.c:149]: Done

**Comment by rschu1ze** [comment]
I sent a CLA as [described](https://openssl.org/policies/cla.html) on the website to [legal@opensslfoundation.org](mailto:legal@opensslfoundation.org)

**Comment by nhorman** [comment]
close/reopen PR to pick up the new cla

**Comment by nhorman** [comment]
hmm, seems that the cla db hasn't been updated yet, will try again soon

File: ssl/ssl_local.h [hunk 3]
```diff
@@ -574,6 +569,12 @@ struct ssl_session_st {
     size_t ticket_appdata_len;
     uint32_t flags;
     SSL_CTX *owner;
+
+    /*
+     * These are used to make removal of session-ids more efficient and to
+     * implement a maximum cache size.
+     */
+    struct ssl_session_st *prev, *next;
```
Comment (mattcaswell) [inline ssl/ssl_local.h:577]: As mentioned please also move the "references" field to after this declaration.

**Review by Sashan — APPROVED** [review]
looks good to me, thank you for taking care of it.

**Comment by rschu1ze** [comment]
Thanks.

Re CLA: I am letting the CCLA check internally, but I don't think there are obstacles to signing it. Once that is done, I'll sign it and send it by mail (I sent the ICLA already yesterday).

**Review by mattcaswell — APPROVED** [review]
LGTM - thanks for the fix

**Review by t8m — APPROVED** [review]

**Comment by rschu1ze** [comment]
Closing + reopening to pick up the C-CLA.

**Comment by nhorman** [comment]
Merged to master and 3.2, thank you for your contribution

note: changes do not apply cleanly to 3.1 and 3.0, please open a new PR to merge commits there

**Comment by t8m** [comment]
I've cherry-picked the commits also to the 3.3 branch. @nhorman it seems you've forgot about that one.

**Comment by nhorman** [comment]
Crap, I did, thank you

**Pull request closed without merging**